In [1]:
from custom_functions import *
from LyapunovQR import *
import numpy as np
import matplotlib.pyplot as plt

In [17]:
def henon(x_arr, alpha, beta):
    #x_arr is a numpy array
    #x_arr[] = [x, y]
    xplus1 = 1 - alpha*x_arr[0]**2 + x_arr[1]
    yplus1 = beta*x_arr[0]
    return np.array([xplus1, yplus1])


#Map M to be used for Lyapunov QR function
def M(x_arr):
    alpha = 0.8
    beta = 0.4
    return henon(x_arr, alpha, beta)

## Lyapunov exponents for Henon Map

In [18]:
N = 100
xini = np.array([1,1.])

lyapunov_exp = LyapunovQR(M, xini, N)
print(lyapunov_exp)

[-0.12522794 -0.78189989]


#### Both Lyapunov Exponents are negative (for $\alpha = 0.8$ and $\beta = 0.4$) implying a non-chaotic system.
#### For $\alpha = 1.4$ and $\beta = 0.3$ we get one positive and one negative lyapunov exponent: [0.409 -1.601]

## Compare to the eigan values of the jacobian.

In [23]:
# find the jacobian at each point
def df(x_arr):
    f = M
    h = 1e-6

    J = MyJacobian(M,x_arr,h)
    J = np.squeeze(J, axis = 2)
    return J


xini = np.array([1,1.])
x = np.zeros((N, n))
n = xini.size #n is the dim we are working in, i.e. M: R^n --> R^n
x[0] = xini


A = np.zeros((N, n, n))
A[0] = df(x[0]) 

i = 0 
while i < N-1:
    x[i+1] = M(x[i])
    A[i+1] = df(x[i+1]) 
    i+=1


In [34]:

print(np.exp(2*lyapunov_exp))
eigval, eigvec = np.linalg.eig(A[2])
print(abs(eigval))

[0.77844583 0.20933912]
[0.8612443 0.4644443]


#### The eigen values of $M_2(p_1)$ are (0.861, 0.464), which are close to  $e^{2\lambda_k} = (0.778,  0.209)$ are similar but not the same.